react/src/util/opReturn.ts

```diff
@@ -82,24 +95,29 @@ function getDataPushdata(data: string, disablePaymentId=false) {
 // `68656c6c6f20776f726c64` - data payload, 'hello world' as ASCII encoded to hex
 // `08` - pushdata for the optional paymentId (paymentId), signifying this tx has 8 bytes of paymentId data
 // `0102030405060708` - The 8-byte paymentId (paymentId)
+//
+// Result: 0450415900001668656c6c6f20776f726c64080102030405060708
 
-export function parseOpReturnProps(
+export function encodeOpReturnProps(
   opReturn: string | undefined,
-  disablePaymentId=false
+  disablePaymentId = false,
```

Comment:
this param is super confusing

we have a negative value that is by default negative

So by default, we have a payment id.

Also confusing, is `opReturn` here the data push?

These params should be matched with the spec, `encodePayButtonTx(dataPush=false, nonce=false)`

Calling `encodePayButtonTx()` should give `0450415900000000` and there should be a unit test for this.

Reply:
I'm open to reversing the negatively-named options but if we do that then we need to be consistent about it and update the others which do in another PR.

The reason for the negatively-named options in the first place is because it's probably easier for people who aren't developers.

Reply:
Yes, since having the `paymentId` is the default, an user will probably look into this parameter (disablePaymentId) when he thinks "how can I get rid of that thing", so having a "disablePaymentId" makes it super clear that that is the option they were looking for.

`opReturn`  is indeed the NULL DATA script dataPush but I decided calling it `opReturn` because it was already what was being done in the telegram channels/foruns/twitter I looked into, as a sort of metonymy for 'the datapush that follows an op_return op code'.  Wouldn't be complicated to rename it all (here and in the server) to `dataPush`, but that definetively would have to be done in a separate issue.

That is also why the function name as of rn is `encodeOpReturnProps`, because it encodes the props `op-return` that the user has passed.  I think calling it `encodePayButtonTx` is not precise, since we are not encoding any tx data besides the `opReturn` (the dataPush, in more correct terms); and we also do encode Txs in another way in the server (using the tx data + a private key to generate a hmac that the user could use to confirm that the POST request about the tx really came from us, and is not spoofed).

As for the other comments, regarding the unittests, I'll make sure to fix all the issues you pointed out and also address it so that their names are more clear about their description and that they match one-to-one the different test cases for what their respective functions propose. 

And finally, regarding the space separated values, indeed it currently will not even accept a syntax like `myKey='quoted string with spaces'`; so that would be a problem. We also support JSON, which would cover this case, but in a more byte-costly way.

So I think we should actually discuss the specs for this key=value parsing feature more thoroughly and arrive at something like the OP_RETURN data push specs, before moving forward with it.

Reply:
Well we talked about supporting spaces with the "simple" PayaButton parsing and decided to just not support it for now, either requiring them to use JSON or supporting it later when there's actually some proven demand for it.

There's really no issue with just doing it later because the on-chain data would be the same. The only difference would be on the server where the submitted OP_RETURN would be updated from just a simple string (where we just give up trying to parse it) to an object that we were able to parse.

Reply:
> There's really no issue with just doing it later because the on-chain data would be the same.

Ideally PayButton won't have to parse all kinds of different data. Option is available for PayButton to control own destiny here and avoid this issue. 

Would be nice if this function is

1 - The only function creating v0 paybutton tx op_return outputs
2 - It only creates them in one format that can always be parsed for v0

That way if you can still change the format by changing the version  number. but you don't need to parse v0 txs in multiple different ways. Is it JSON? Is it utf8 string? What if there's another idea and it's very tempting to change it without changing the version?

It's all fixable but I could see it causing a lot more work for PayButton down the road.

Reply:
> That is also why the function name as of rn is encodeOpReturnProps, because it encodes the props op-return that the user has passed. I think calling it encodePayButtonTx is not precise, 

well -- the spec is for v0 paybutton txs, which is the only version this function handles. At least document it in the function comments.

If there is ambiguity into what is or is not a v0 paybutton tx, should be addressed in the spec. we're assuming that the tx is a valid ecash tx -- what makes it a paybutton v0 tx is the op_return output conforming to the documented spec

Reply:
> Would be nice if this function is the only function creating v0 paybutton tx op_return outputs

It should be - I'm not sure if things will get more complicated if we're eventually doing BIP 70 but I'm pretty sure we will still be able to keep it just within the button.

> It only creates them in one format that can always be parsed for v0

1. We do want to at least allow entering simple bits of text as simply entering JSON isn't practical for much of the target audience (who are often not developers).
2. The spec for the "data" portion is just "here's a bucket of data". If we can successfully parse it into JSON on the server, that's good but if not it's not a big deal. We can't guarantee good data in the OP_RETURN anyway since it's permissionless -- also typos are a thing.
3. Every app using PayButton is going to be different. Some people will have weird (to us) systems for managing things. We can encourage people to do stuff a certain way but limiting them when they really really want to do it some other way is probably a bad idea in this case. I'm reminded of CoinMarketCap and how they explicitly requested that the API we built for them return "just the number" because they couldn't handle parsing JSON (which is what we provided them with originally). So in our case here, they'll either be getting back a JSON object or a simple string in their payment trigger (server webhook) if they have that set up -- and they can do whatever they want with it at that point.

If a 3rd party service wanted to go through and index paybutton txs, they should be able to follow the spec, just treating the "data" portion as a big string.

Reply:
The idea of trying to parse 2 things is mainly to give people options who find JSON scary -- which might sound like a dumb reason but it's a real problem for people who aren't developers but who will want to do things like sorting, filtering, etc of transactions through the PayButton (server) UI.

I think it could be worth thinking about doing the JSON parsing attempt on the client button instead of having the server do it, but there's the tradeoff of JSON taking up more space. Still worth discussing.

Reply:
> We do want to at least allow entering simple bits of text as simply entering JSON isn't practical for much of the target audience (who are often not developers).

imo it's risky having this -- why not condition all the input data to meet your spec? if json is intimidating, can have multiple 'key' and 'value' fields, for example. Even without this, could still have some sort of preprocessing between what the user is entering and what is encoded in the tx.

Or, even simpler things, like "strings cannot have spaces", or "replace all spaces in strings with hex char for space"

Reply:
We could use a query parser such as  `qs` and `JSON.parse`. So we just try both and return whatever works, or a string. I don't think they ever overlap, and if they do, will be probably in edge cases in which we will just return an empty string anyway.

This also would be easy to document, since it already is.

Reply:
Actually sounds like a decent idea to me, rather than rolling our own and attempting to fix it up later.

@BytesOfMan thoughts?

Reply:
Yeah sounds like a good way to validate the user input. Only accept user input that passes the parser you want to use.

it's not really a spec thing, but you want some kind of limitation on what type/shape of data gets into the tx.